aibolit/patterns/mutable_index/mutable_index.py

```diff
@@ -2,21 +2,66 @@
 # SPDX-License-Identifier: MIT
 import os
 
+from aibolit.ast_framework.ast import AST
+from aibolit.ast_framework.ast_node import ASTNode
+from aibolit.ast_framework.ast_node_type import ASTNodeType
+from aibolit.types_decl import LineNumber
+from aibolit.utils.ast_builder import build_ast
 
-class MutableIndex:
 
+class MutableIndex:
     def __init__(self):
         pass
 
-    def value(self, filename: str | os.PathLike):
+    def value(self, filename: str | os.PathLike) -> list[LineNumber]:
         """
         Traverse over AST tree and finds loops with mutable index
 
         :return: List of line number of loops with mutable index
-
-        @todo #147:30min Implement MutableIndex pattern
-        We need to implement the code on mutable index pattern. It must return the number
-        of the lines where we have an index of a for loop being mutated. After implementing,
-        the method, enable all tests in test_mutable_index.py
         """
-        return []
+        result = set()
+        ast = AST.build_from_javalang(build_ast(filename))
+        for node in ast.get_proxy_nodes(
+            ASTNodeType.FOR_STATEMENT,
+        ):
+            index_names = self._collect_index_names(ast, node)
+            for_body = ast.get_subtree(node.body)
+            result.update(self._handle_assignment(index_names, for_body))
+            result.update(self._handle_unary_operations(index_names, for_body))
```

Comment:
@AntonProkopyev I appreciate the way you factored out helper methods here for readability!